server/app/routes/v1/volumes_api.rb

```diff
@@ -0,0 +1,113 @@
+module V1
+  class VolumesApi < Roda
+    include TokenAuthenticationHelper
+    include CurrentUser
+    include RequestHelpers
+    include Auditor
+    include LogsHelpers
+
+    plugin :backtracking_array
+
+    route do |r|
+
+      validate_access_token
+      require_current_user
+
+      def find_grid(grid_name)
+        grid = Grid.find_by(name: grid_name)
+        halt_request(404, {error: "Grid #{grid_name} not found"}) if !grid
+        grid
+      end
+
+      def find_stack(stack_name, grid)
+        stack = grid.stacks.find_by(name: stack_name)
+        halt_request(404, {error: "Stack #{stack_name} not found"}) if !stack
+        stack
+      end
+
+      # @param [String] grid_name
+      # @param [String] stack_name
+      # @param [String] volume_name
+      # @return [GridService]
+      def load_volume(grid_name, stack_name, volume_name)
+        grid = find_grid(grid_name)
+        if stack_name
+          stack = find_stack(stack_name, grid)
+          volume = stack.volumes.find_by(name: volume_name)
+        else
+          volume = grid.volumes.find_by(name: volume_name)
+        end
+        halt_request(404, {error: "Volume #{volume_name} not found"}) if !volume
+
+        unless current_user.grid_ids.include?(volume.grid_id)
+          halt_request(403, {error: 'Access denied'})
+        end
+
+        volume
+      end
+
+      # Parse stack/vol out of path
+      # @param [String] volume path
+      # @return [Array] [stack, volume]
+      def parse_params(path)
```

Comment:
Agreed that the `null` in the `/v1/volumes/:grid/null/:volume` URL is a bit odd, and I would also leave the special `null` stack out of this, the current `return [nil, volume]` is much better.

But if volumes are going to be first-class managed resources, then they might grow extra sub-resources at some point in the future, like say snapshots... if you look at how those work in e.g. the services `/v1/services/testgrid/null/fooservice/containers` API, then this kind of variable-length URL scheme would not work.

If we don't need to be consistent with the services API, then maybe even try `/v1/volumes/:grid/:stack.:volume` vs `/v1/volume/:grid/:volume`, like the Docker container names? But maybe it's better to just be consistent with what we already have...